scripts/harbor-init/harbor-init.go

```diff
@@ -0,0 +1,236 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+)
+
+const (
+	// Harbor API version
+	HARBOR_API_ENDPOINT = "/api/v2.0"
+)
+
+type Login struct {
+	Url      string
+	Username string
+	Password string
+}
+
+type Projects []struct {
+	Name string `json:"name"`
+}
+
+type ProjectRobots []struct {
+	Name        string `json:"name"`
+	Id          int    `json:"id"`
+	Permissions []struct {
+		Namespace string `json:"namespace"`
+	} `json:"permissions"`
+}
+
+type ResponseRobot struct {
+	Id int `json:"id"`
+}
+
+func request(method string, login Login, endpoint string, body []byte) (*http.Response, []byte, error) {
+	client := &http.Client{}
+	req, err := http.NewRequest(method, login.Url+endpoint, bytes.NewBuffer(body))
+	if err != nil {
+		return nil, nil, err
+	}
+	req.SetBasicAuth(login.Username, login.Password)
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer resp.Body.Close()
+	respBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, nil, err
+	}
+	return resp, respBody, nil
+}
+
+func main() {
+	urlFlag := flag.String("url", "https://harbor.example.com", "Harbor URL")
+	adminFlag := flag.String("admin", "admin", "Harbor admin username")
+	passwordFlag := flag.String("password", "", "Harbor admin password")
+	projectNameFlag := flag.String("project", "renku-build", "Project name")
+	robotAccountNameFlag := flag.String("robot", "renku-registry-robot", "Robot account name")
+	robotSecretFlag := flag.String("secret", "", "Robot account secret")
+	flag.Parse()
+	// Initialize logger to print to stdout
+	logger := log.New(os.Stdout, "INFO: ", log.Ldate|log.Ltime)
+	login := Login{}
+	// Get login from environment variable
+	if *urlFlag == "" {
+		logger.Fatal("url flag is not set")
+	}
+	login.Url = *urlFlag + HARBOR_API_ENDPOINT
+	login.Username = *adminFlag
+	if login.Username == "" {
+		logger.Fatal("admin flag is not set")
+	}
+	login.Password = *passwordFlag
+	if login.Password == "" {
+		logger.Fatal("password flag is not set")
+	}
+	projectName := *projectNameFlag
+	if projectName == "" {
+		logger.Fatal("project flag is not set")
+	}
+	robotAccountName := *robotAccountNameFlag
+	if robotAccountName == "" {
+		logger.Fatal("robot flag is not set")
+	}
+	robotSecret := *robotSecretFlag
+	if robotSecret == "" {
+		logger.Fatal("secret flag is not set")
+	}
+
+	// Authenticate and get a session cookie
+	logger.Println("Authenticating with Harbor server")
+	resp, body, err := request("GET", login, "/audit-logs", nil)
+	if err != nil {
+		logger.Fatal("Error connecting to Harbor: ", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		logger.Fatal("Failed to authenticate:", resp.Status, err, string(body))
+		return
+	}
+	logger.Println("Authenticated successfully")
```

Comment:
question: It may be that this is not needed if you send the credentials in basic auth header every time with the request method? Or if you want to keep cokies around then probabl you dont want to initialize http.Client from scratch on every call. I am pretty sure that wipes the cookies. Or you may need to add some kind of cookie jar provider to the http client if you want to reuse cookies.

Reply:
the client is now init in the main and passed around with the `login` type